go-controller/pkg/node/gateway_init.go

```diff
@@ -82,15 +82,84 @@ func getIPv4Address(iface string) (*net.IPNet, error) {
 	if err != nil {
 		return nil, err
 	}
+	var ips []*net.IPNet
+	var foundIPv4 bool
+	var foundIPv6 bool
 	for _, addr := range addrs {
 		switch ip := addr.(type) {
 		case *net.IPNet:
-			if !utilnet.IsIPv6CIDR(ip) {
-				return ip, nil
+			if utilnet.IsIPv6CIDR(ip) {
+				if config.IPv6Mode && !foundIPv6 {
+					ips = append(ips, ip)
+					foundIPv6 = true
+				}
+			} else if config.IPv4Mode && !foundIPv4 {
+				ips = append(ips, ip)
+				foundIPv4 = true
 			}
 		}
 	}
-	return nil, nil
+	if config.IPv4Mode && !foundIPv4 {
+		return nil, fmt.Errorf("failed to find IPv4 address on interface %s", iface)
+	} else if config.IPv6Mode && !foundIPv6 {
+		return nil, fmt.Errorf("failed to find IPv6 address on interface %s", iface)
+	}
+	return ips, nil
+}
+
+func getGatewayNextHops() ([]net.IP, string, error) {
+	var gatewayNextHops []net.IP
+	var needIPv4NextHop bool
+	var needIPv6NextHop bool
+
+	if config.IPv4Mode {
+		needIPv4NextHop = true
+	}
+	if config.IPv6Mode {
+		needIPv6NextHop = true
+	}
+
+	// FIXME DUAL-STACK: config.Gateway.NextHop should be a slice of nexthops
+	if config.Gateway.NextHop != "" {
+		// Parse NextHop to make sure it is valid before using. Return error if not valid.
+		nextHop := net.ParseIP(config.Gateway.NextHop)
+		if nextHop == nil {
+			return nil, "", fmt.Errorf("failed to parse configured next-hop: %s", config.Gateway.NextHop)
+		}
+		if config.IPv4Mode && !utilnet.IsIPv6(nextHop) {
+			gatewayNextHops = append(gatewayNextHops, nextHop)
+			needIPv4NextHop = false
+		}
+		if config.IPv6Mode && utilnet.IsIPv6(nextHop) {
+			gatewayNextHops = append(gatewayNextHops, nextHop)
+			needIPv6NextHop = false
+		}
+	}
+	gatewayIntf := config.Gateway.Interface
+	if needIPv4NextHop || needIPv6NextHop || gatewayIntf == "" {
```

Comment:
this if statement doesn't make sense to me. if you need ipv4 or ipv6 next hops, but gatewayIntf is provided, you shouldn't just do getDefaultGatewayInterfaceDetails right? That will search all routes instead of just the ones for the gateway interface. It seems like it should be:

```
if needIPv4NextHop || needIPv6NextHop {
         if gatewayIntf == "" {
              defaultGatewayIntf, defaultGatewayNextHops, err := getDefaultGatewayInterfaceDetails()
         } else {
            // find ipv4 and ipv6 routes for gatewayIntf
        }

```

Reply:
This matches the older single-stack logic. The autodetection can only handle simple cases. Of course, given that you can only manually specify a single NextHop value right now, this means you can't currently do dual stack shared gateway mode on a node with multiple default gateways where the one it autodetects isn't the one you wanted. To be revisited.